### ISSUE TITLE
Restrict access to specified folder and child folders only (not allow user to browse to parent folders)

### ISSUE THREAD

**Issue by gordon-matt** [trippo/ResponsiveFilemanager#294]
Is it possible to restrict the users to their own root folders? I have 2 different situations:

1) I am implementing a "Tenants" feature for my CMS, so multiple sites can run on the same instance. However, this means they must not share the same media folders. I would like to have something like this:
- Root
   - Tenant 1
   - Tenant 2
   - etc

I can pass "fldr" parameter and that is a good start, but the users can still use the "home icon" to browse to the root folder and then access files for the other tenants. This is not good.

2) The second situation is mostly the same as the first one, but for individual users... so example sturcture I would like my CMS to have is:

- Root
   - Tenant 1
      - User 1
      - User 2
      - etc
   - Tenant 2
      - User 3
      - etc
   - etc

I think it would be good to have a new parameter like: "rootFolder" or something, so it will be same as "fldr" but will restrict the users so they cannot browse to parent folders.. and the home icon will redirect make to that specified root folder. The site security is very important, so we don't have users accessing files that they should not.

**Comment by cdaveb** [comment]
We have something similar to this implemented on our company's internal 
admin site- users have access to different folders and we have a custom 
function that sets the hidden directories based on user permissions 
(we've tweaked our config to have access to the relevant user session 
information). It's a bit of a hack but it does the job.

It would be nice if there was a more standard way of doing it.

**Comment by trippo** [comment]
You can read SET SUBFOLDER AS ROOT AND MULTI-USER USE section on documentation

**Comment by gordon-matt** [comment]
@trippo , thanks for the reply. You probably forgot I am the guy using it in .NET. I wrote the article on CodeProject, "[Using Responsive File Manager in ASP.NET MVC](http://www.codeproject.com/Tips/1004596/Using-Responsive-File-Manager-in-ASP-NET-MVC)". So I am not sure how to set the `$_SESSION["RF"]["subfolder"]` variable in my case. Is it possible for me to do the same thing some other way? Otherwise, maybe I need to try modify your code to pass a new parameter in the URL and set that `$_SESSION["RF"]["subfolder"]` variable inside dialog.php? I guess that's what I have to do, but I'm not sure because I don't know much about PHP.

**Comment by trippo** [comment]
Yes you must edit dialog.php in the top of file where read $_SESSION['RF']["subfolder"] and change to read $_GET['subfolder']

But is not secure because an user can see this in url and can change

**Comment by gordon-matt** [comment]
@trippo , Yes, that is unfortunate. I am very worried about security. However, I don't see any other way to do this. Can you make some kind of option in the config file to use a javascript function to call specified ajax url to get the root directory for a user? Because that way, I can get the correct tenant and user info and return a string for the correct root folder URL. Example, we can put something like this in config file:

**getRootFolderUrl = '/url-to-my-ajax-function'**

and then your dialog.php can use that to query for correct root folder each time.. is it possible?

Last thing: are you still working on the new JavasScript-only version (Angular) ? If yes, do you have any estimated timeline for when that will be released?

**Comment by gordon-matt** [comment]
OK I can't figure out how to set a $_SESSION variable to some value from an ajax request. So for now I changed some code under the ***SUB-DIR CODE** section in dialog.php:

From this:

```
if (!isset($_SESSION['RF']["subfolder"]))
{
	$_SESSION['RF']["subfolder"] = '';
}
```

to this:

```
if (isset($_GET['rootFolder']))
{
	$_SESSION['RF']["subfolder"] = $_GET["rootFolder"];
}
else if (!isset($_SESSION['RF']["subfolder"]))
{
	$_SESSION['RF']["subfolder"] = '';
}
```

and then my href on the iframe becomes something like:

`<iframe src="/filemanager/dialog.php?type=0&rootFolder=Tenant2".....`

This is not secure, but it works. It would be awesome if you could make a fix in the next version to set some ajax url in the config file and then the dialog.php will call that url to get the correct root folder each time.

**Comment by gordon-matt** [comment]
Also, how can I pass a parameter to tinymce? I need to do the same thing for that...

**Comment by cdaveb** [comment]
What we ended up doing for our custom configuration was comment out the 
session start and initialized the session via our custom code so that it 
would have access to the session information we needed to determine 
which folders were appropriate for the user. This would require you 
writing a bit of PHP, but you might be able to do something similar and 
it's a more secure mechanism as you're getting the information you want 
from the server.

You could also use a cookie to get the information if you set one when 
the user logs in- that also isn't that secure but at least it wouldn't 
be visible in the URL and as easily manipulated.

**Comment by raknjak** [comment]
@cdaveb do you have an example available please?

**Comment by cdaveb** [comment]
Our specific code wouldn't be a very good example as it involves including files from our internal PHP libraries. The main thing is you'd just want to start the session via whatever means you normally do in your application by triggering it through the code at the beginning of config.php- then you can read from whatever session variables are stored by your application. If your application is not in PHP then I'm not sure how workable that would be, but it might still be doable depending on how your session information is stored.

If you want to use the cookie approach, you'd just need to make sure that when a user logs in it stores a cookie with the appropriate information you need to display the correct subfolder. As I said this would be less secure as the user could manually change the cookie if they knew what they were doing, but it's a little less obvious than having it in the URL.
